### PR TITLE
RFC: advice needed regarding integration testing with in memory mongo

### PR DESCRIPTION
**Pull request by esclapes**
Heya @hharnisc @djfarrelly and @philippemiguet,

I would like to get your thoughts/advice regarding this spike.

During Smokey I added a new RPC method to this service and I felt a bit insecure mocking the model. I have little experience with MongoDB and mongoose, so I was not sure if the model would behave the way I was mocking it.

In the past, I have used in-memory DBs to avoid mocking the ORM models, and reading about testing Mongo with Jest, I found [this same approach was the suggested](https://jestjs.io/docs/en/mongodb).

In this PR you can find an example of how it would impact this service codebase to introduce an in-memory mongoDB for integration testing.

The idea is simple. This service stays a pure express app (mongo does not need to be installed in the docker image). When running tests, a specific mongodb binary is downloaded (only on first run) via the `mongodb-memory-server` package. A new mongo instance is initialized with in-memory persistence and a connection string is provided for mongoose to connect during the tests.

I can see a few benefits of this approach:

* This service relies heavily on mongoose models, which in turn relies on mongoDB. Both of them are hard to mock. With an in-memory db we do not need to mock them so there is less chance of false positives.
* Tests become a bit simpler: set initial state, call method, check new state of exception.
* As the RPC unit tests require less mocking, they become less coupled to the implementation. We could refactor an update using `save` to an update using `updateOne` without having to change the mocking in our tests.
* I see this service is mainly a _gatekeeper_ to the data in mongoDB. Testing it with a real mongoDB instance allows us to ensure the expected final state after any method call.
* Mocking is still available for testing edge cases that might not be easy to set up.
* It is opt-in, as the specific jest `test-environment` is defined via annotation on the top of the test file.

Of course, there are trade-offs:

* This tests are going to be slower. I am hoping the in-memory version is fast enough for our use-case and I am happy to rewrite all current tests to really asses the impact here.
* This kind of integration testing may be less detailed when pointing at the root cause: which piece of code is broken. Still the service is thin enough that this might not be a big issue.
* There is a bit of setup required for this tests: `beforeAll`, `afterAll` and `beforeEach`. I tried to extract it into `mongo-environemnt.js`, but somehow the model is not wired correctly. I still think it should be possible. Did not want to spend too much time on this.

**The biggest issue** (and blocker) is that the `mongodb-memory-server` **cannot find and download the right binary for an Alpine container,** because there is none hosted on the official MongoDB servers. This means that this tests can only be run locally (outside of the container). Running the tests in the laptop should be fine as any other service would be mocked, but it might be an issue if running in CI. I can think of these two solutions:

* Build (or get) our own binary and `COPY` it to the expected path into the container. It will only be used during tests. It will increase the size of the container. For reference the ubuntu binary weights 57Mb
* Move to use an `node:jessie` image which is supported out of the box
 
Looking forward to hearing your thoughts and advice on this.

### CODE REVIEW

File: tests/rpc/createAccount.test.js [hunk 1]
```diff
@@ -1,74 +1,93 @@
-jest.mock('../../src/models/authenticationAccountModel')
+/**
+ * @jest-environment ./tests/mongo-environment.js
```
Comment (hharnisc) [inline tests/rpc/createAccount.test.js]: can we put this in the jest config object?

File: tests/mongo-environment.js [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+module.exports = class MongoEnvironment extends NodeEnvironment {
+  constructor(config) {
+    super(config)
+  }
+
+  async setup() {
+    // console.log('Setup mongo')
+    const server = new MongoMemoryServer({
+      binary: {
+        version: '3.6.7',
```
Comment (hharnisc) [inline tests/mongo-environment.js]: Glad we can lock down the version :smile:

File: tests/mongo-environment.js [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+    this.global.__MONGO__ = {
+      server,
+      uri: await server.getConnectionString(),
+    }
+
+    await super.setup()
+  }
+
+  async teardown() {
+    // console.log('Teardown mongo')
```
Comment (hharnisc) [inline tests/mongo-environment.js]: looks like a left over comment :pencil2:

File: tests/mongo-environment.js [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+    await this.global.__MONGO__.server.stop()
+
+    await super.teardown()
+  }
+
+  runScript(script) {
+    return super.runScript(script)
```
Comment (hharnisc) [inline tests/mongo-environment.js]: looks like we don't need this class function :smile:

File: tests/mongo-environment.js [hunk 5]
```diff
@@ -0,0 +1,36 @@
+const NodeEnvironment = require('jest-environment-node')
+const { MongoMemoryServer } = require('mongodb-memory-server')
+
+module.exports = class MongoEnvironment extends NodeEnvironment {
+  constructor(config) {
+    super(config)
```
Comment (hharnisc) [inline tests/mongo-environment.js]: I don't remember this one off the top of my head - but since constructor only calls super we might be able to omit it.

**Comment by esclapes** [comment]
Good question @djfarrelly!

I would still prefer this approach over mocking because I think testing with an in-memory MongoDB gives better warranties and more leverage. For instance, it would allow for easy automated testing of the migration script.

I recently came back to this PR and tried to copy the required binary inside the `alpine` container, but there were some shared libraries missing and it did not work, so **I am thinking the easy way to unblock this is moving from `alpine` to `slim` image which is ~250Mb  (5 times the size of `alpine`)**. 

How do you feel about that possibility? Would that increase in image size have impact at the infra level? Do you foresee any other trade-offs?

**Comment by esclapes** [comment]
Hey @djfarrelly I just realized that we run our tests in Travis CI outside of the docker container, and their default build environment is Ubuntu Linux. That means that the `mongodb-memory-server` would just work.

We could then use the `-slim` base image for development to be able to run the tests locally, and the `-alpine` for production to reduce the image size. 

I wanted to share this here as it feels like one less blocker. However, I don't want to jump to the technicalities before we have agreed on this change being worth it. 

I would love to know how this idea feels to you?

**Comment by djfarrelly** [comment]
Those sound like great ideas to me @esclapes, I'd definitely aim to run alpine in production and ideally it would be great to keep the same base container image in local development and in production, but I guess for improved testing that's alright. You're totally right that Travis uses Ubuntu which should make this much simpler for CI 😄 

I'll leave this decision on how to proceed to you! Happy with whatever path you choose 👍

**Comment by esclapes** [comment]
Thanks @djfarrelly  for all the advice. 

I am going to close this PR as I have re-used part of the code and insight from this thread in the upcoming #66
